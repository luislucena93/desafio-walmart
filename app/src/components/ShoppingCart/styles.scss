@import '../../commons/styleVars';

.overlay {
  padding-top: 100px;
  position: absolute;
  width: 0;
  height: 0;
  background-color: rgba(0, 0, 0, 0);
  transition: background-color 0.3s ease-in;
  &.show {
    background-color: rgba(0, 0, 0, 0.5);
    width: 100%;
    height: 100%;
    .cartContent {
      visibility: visible;
    }
    .cart {
      visibility: visible;
    }
  }
}

.cart {
  width: 400px;
  background-color: white;
  position: absolute;
  right: 0px;
  transform: translateX(400px);
  transition: transform 0.3s ease-in;
  visibility: hidden;
  &.open {
    transform: translateX(0px);
  }
}

.cartContent {
  color: $darkGray;
  visibility: hidden;
  margin: 20px 10px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  .title {
    margin-top: 0px;
    color: $mainColor;
    font-weight: 600;
    font-size: 18px;
    padding-bottom: 10px;
    border-bottom: 1px solid $shadowColor;
  }
  .selectedProductsList {
    background-color: $backgroundColor;
    border: solid 1px $shadowColor;
    width: 100%;
    max-height: 350px;
    overflow-y: scroll;
    flex-grow: 3;
  }
}

.discuountSuggestion {
  color: $mainColor;
  font-size: 12px;
  margin-top: 5px;
  span {
    font-weight: 600;
  }
}

.labelAmount {
  display: flex;
  justify-content: space-between;
  &.appliedDiscount {
    p {
      margin-bottom: 0px;
    }
  }
  &.total {
    font-size: 14px;
    font-weight: 600;
  }
}

.discountText {
  margin-top: 5px;
  font-size: 12px;
  span {
    font-weight: 600;
  }
}

.bold {
  font-weight: 600;
}