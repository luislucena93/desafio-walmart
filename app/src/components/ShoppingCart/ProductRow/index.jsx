import React from 'react';
import './styles.scss';
import FormatMoney from '../../../helpers/formatters';

const ProductRow = (props) => {
    const {product, removeProduct, addProduct} = props;
    return (
        <div className={"productRow"}>
            <img src={`http://${product.info.image}`}/>
            <div className={"productSummary"}>
                <p className={"bold"}>{product.info.brand}</p>
                <p>{product.info.description}</p>
                <p>${FormatMoney(product.info.price)}</p>
                <p className={"bold"}>Cantidad: {product.count}</p>
            </div>
            <div className={"productTotal"}>
                <p className={"bold"}>${FormatMoney(product.count * product.info.price)}</p>
                <div className={"addRemoveButtons"}>
                    <button onClick={() => removeProduct(product.info)}>-</button>
                    <p>{product.count}</p>
                    <button onClick={() => addProduct(product.info)}>+</button>
                </div>
            </div>
        </div>
    )
};

export default ProductRow;