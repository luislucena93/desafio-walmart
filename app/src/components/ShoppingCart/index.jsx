import React, {useState, useEffect} from 'react';
import './styles.scss';
import ProductRow from './ProductRow/index.jsx';
import FormatMoney from '../../helpers/formatters';

const ShoppingCart = (props) => {
    const {show, closeShoppingCart, selectedProducts, brandTotals, discounts, addProduct, removeProduct} = props;
    const [open, setOpen] = useState(false);
    let subtotal = 0;
    Object.keys(selectedProducts).forEach((key) => {
        subtotal += selectedProducts[key].count * selectedProducts[key].info.price;
    });

    useEffect(() => {
        if (show) {
            setOpen(true);
        }
    }, [show]);

    const waitToClose = (e) => {
        if (e.target !== e.currentTarget) return;
        setOpen(false);
        setTimeout(() => {
            closeShoppingCart()
        }, 500);
    };

    const getDiscountSuggestion = () => {
        if (discounts.maxDiscountPossible && (!discounts.maxDiscountReached || discounts.maxDiscountReached.brand !== discounts.maxDiscountPossible.brand)) {
            return (
                <p className={"discuountSuggestion"}>
                    Agrega&nbsp;
                    <span>${FormatMoney(discounts.maxDiscountPossible.threshold - brandTotals[discounts.maxDiscountPossible.brand])}</span>&nbsp;
                    mas en productos {discounts.maxDiscountPossible.brand} y aprovecha un descuento total de&nbsp;
                    <span>${FormatMoney(discounts.maxDiscountPossible.discount)}</span> en tu compra!
                </p>
            )
        }
        return null
    };

    const getProductsSubtotal = () => {
        return (
            <div className={"labelAmount"}>
                <p>Subtotal de productos</p>
                <p>${FormatMoney(subtotal)}</p>
            </div>
        )
    };

    const getProductsTotal = () => {
        const total = discounts.maxDiscountReached ? subtotal - discounts.maxDiscountReached.discount : subtotal;
        return (
            <div className={"labelAmount total"}>
                <p>Total a pagar</p>
                <p>${FormatMoney(total)}</p>
            </div>
        )
    };

    const getAppliedDiscount = () => {
        if (discounts.maxDiscountReached) {
            return (
                <div>
                    <div className={"labelAmount appliedDiscount"}>
                        <p>Descuento por marca</p>
                        <p>-${FormatMoney(discounts.maxDiscountReached.discount)}</p>
                    </div>
                    <p className={"discountText"}>
                        Se aplic√≥ un descuento de <span>${FormatMoney(discounts.maxDiscountReached.discount)}</span> por
                        haber&nbsp;
                        comprado al menos <span>${FormatMoney(discounts.maxDiscountReached.threshold)}</span> en&nbsp;
                        productos {discounts.maxDiscountReached.brand}
                    </p>
                </div>
            )
        }
        return null;
    };

    return (
        <div className={`overlay ${show ? 'show' : ''}`} onClick={waitToClose} data-testid={'overlay'}>
            <div className={`cart ${open ? 'open' : ''}`} id={"cart"} data-testid={'cart'}>
                <div className={"cartContent"}>
                    <p className={"title"}>Resumen</p>
                    <div className={"selectedProductsList"}>
                        {
                            Object.keys(selectedProducts).map((key, i) => <ProductRow key={i}
                                                                                      product={selectedProducts[key]}
                                                                                      addProduct={addProduct}
                                                                                      removeProduct={removeProduct}/>)
                        }
                    </div>
                    <div className={"totals"}>
                        {getDiscountSuggestion()}
                        {getProductsSubtotal()}
                        {getAppliedDiscount()}
                        {getProductsTotal()}
                    </div>

                </div>
            </div>
        </div>
    )
}

export default ShoppingCart;