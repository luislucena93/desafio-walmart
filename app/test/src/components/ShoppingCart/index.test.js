import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react';
import ShoppingCart from "../../../../../app/src/components/ShoppingCart/index";

describe('ShoppingCart component', () => {

    let mockCloseShoppingCart;
    let mockAddProduct;
    let mockRemoveProduct;
    const testSelectedProduct = {
        testId: {
            count: 2,
            info: {id: 'testId', brand: 'testBrand', description: 'testDescription', image: 'testImage', price: 1000}
        }
    };
    const testBrandTotals = {testBrand: 1000};
    let testDiscounts;

    beforeEach(() => {
        mockCloseShoppingCart = jest.fn();
        mockAddProduct = jest.fn();
        mockRemoveProduct = jest.fn();
        testDiscounts = {maxDiscountReached: null, maxDiscountPossible: null};
    });

    it('should render ShoppingCart correctly if show is false', () => {
        const {container} = render(<ShoppingCart show={false} selectedProducts={testSelectedProduct}
                                                 closeShoppingCart={mockCloseShoppingCart} brandTotals={testBrandTotals}
                                                 discounts={testDiscounts}
                                                 addProduct={mockAddProduct}
                                                 removeProduct={mockRemoveProduct}/>);
        expect(container).toMatchSnapshot();
    });

    it('should render ShoppingCart correctly if show is true', () => {
        const {container} = render(<ShoppingCart show={true} selectedProducts={testSelectedProduct}
                                                 closeShoppingCart={mockCloseShoppingCart} brandTotals={testBrandTotals}
                                                 discounts={testDiscounts}
                                                 addProduct={mockAddProduct}
                                                 removeProduct={mockRemoveProduct}/>);
        expect(container).toMatchSnapshot();
    });

    it('should show discount suggestion and not show applied discount if it has maxDiscountPossible and not maxDiscountReached', () => {
        testDiscounts.maxDiscountPossible = {threshold: 10000, discount: 1000, brand: 'testBrand'};
        const {container} = render(<ShoppingCart show={true} selectedProducts={testSelectedProduct}
                                                 closeShoppingCart={mockCloseShoppingCart} brandTotals={testBrandTotals}
                                                 discounts={testDiscounts}
                                                 addProduct={mockAddProduct}
                                                 removeProduct={mockRemoveProduct}/>);
        expect(container).toHaveTextContent(/Agrega \$9.000 mas en productos testBrand y aprovecha un descuento total de \$1.000 en tu compra!/i);
        expect(container).not.toHaveTextContent(/Descuento por marca/i)
    });

    it('should show discount suggestion and applied discount if maxDiscountPossible and maxDiscountReached brand are different', () => {
        testDiscounts.maxDiscountPossible = {threshold: 10000, discount: 1000, brand: 'testBrand'};
        testDiscounts.maxDiscountReached = {threshold: 8000, discount: 800, brand: 'testBrand1'};
        const {container} = render(<ShoppingCart show={true} selectedProducts={testSelectedProduct}
                                                 closeShoppingCart={mockCloseShoppingCart} brandTotals={testBrandTotals}
                                                 discounts={testDiscounts}
                                                 addProduct={mockAddProduct}
                                                 removeProduct={mockRemoveProduct}/>);
        expect(container).toHaveTextContent(/Agrega \$9.000 mas en productos testBrand y aprovecha un descuento total de \$1.000 en tu compra!/i);
        expect(container).toHaveTextContent(/Descuento por marca-\$800/i);
        expect(container).toHaveTextContent(/Se aplicÃ³ un descuento de \$800 por haber comprado al menos \$8.000 en productos testBrand1/i)
    });

    it('should not show discount suggestion if maxDiscountPossible and maxDiscountReached are equal', () => {
        testDiscounts.maxDiscountPossible = {threshold: 10000, discount: 1000, brand: 'testBrand'};
        testDiscounts.maxDiscountReached = {threshold: 10000, discount: 1000, brand: 'testBrand'};
        const {container} = render(<ShoppingCart show={true} selectedProducts={testSelectedProduct}
                                                 closeShoppingCart={mockCloseShoppingCart} brandTotals={testBrandTotals}
                                                 discounts={testDiscounts}
                                                 addProduct={mockAddProduct}
                                                 removeProduct={mockRemoveProduct}/>);
        expect(container).not.toHaveTextContent(/Agrega \$9.000 mas en productos testBrand y aprovecha un descuento total de \$1.000 en tu compra!/i);
    });

    it('should call close shopping cart function after 500ms only when click on overlay outside shopping cart', () => {
        jest.useFakeTimers();
        render(<ShoppingCart show={true} selectedProducts={testSelectedProduct}
                             closeShoppingCart={mockCloseShoppingCart} brandTotals={testBrandTotals}
                             discounts={testDiscounts}
                             addProduct={mockAddProduct}
                             removeProduct={mockRemoveProduct}/>);
        fireEvent.click(screen.getByTestId('cart'));
        expect(mockCloseShoppingCart).not.toHaveBeenCalled();
        fireEvent.click(screen.getByTestId('overlay'));
        jest.runAllTimers();
        expect(mockCloseShoppingCart).toHaveBeenCalled();
        jest.useRealTimers();
    });

});