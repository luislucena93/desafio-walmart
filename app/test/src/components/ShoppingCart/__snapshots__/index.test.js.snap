// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCart component should render ShoppingCart correctly if show is false 1`] = `
<div>
  <div
    class="overlay "
    data-testid="overlay"
  >
    <div
      class="cart "
      data-testid="cart"
      id="cart"
    >
      <div
        class="cartContent"
      >
        <p
          class="title"
        >
          Resumen
        </p>
        <div
          class="selectedProductsList"
        >
          <div
            class="productRow"
          >
            <img
              src="http://testImage"
            />
            <div
              class="productSummary"
            >
              <p
                class="bold"
              >
                testBrand
              </p>
              <p>
                testDescription
              </p>
              <p>
                $
                1.000
              </p>
              <p
                class="bold"
              >
                Cantidad: 
                2
              </p>
            </div>
            <div
              class="productTotal"
            >
              <p
                class="bold"
              >
                $
                2.000
              </p>
              <div
                class="addRemoveButtons"
              >
                <button>
                  -
                </button>
                <p>
                  2
                </p>
                <button>
                  +
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="totals"
        >
          <div
            class="labelAmount"
          >
            <p>
              Subtotal de productos
            </p>
            <p>
              $
              2.000
            </p>
          </div>
          <div
            class="labelAmount total"
          >
            <p>
              Total a pagar
            </p>
            <p>
              $
              2.000
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ShoppingCart component should render ShoppingCart correctly if show is true 1`] = `
<div>
  <div
    class="overlay show"
    data-testid="overlay"
  >
    <div
      class="cart open"
      data-testid="cart"
      id="cart"
    >
      <div
        class="cartContent"
      >
        <p
          class="title"
        >
          Resumen
        </p>
        <div
          class="selectedProductsList"
        >
          <div
            class="productRow"
          >
            <img
              src="http://testImage"
            />
            <div
              class="productSummary"
            >
              <p
                class="bold"
              >
                testBrand
              </p>
              <p>
                testDescription
              </p>
              <p>
                $
                1.000
              </p>
              <p
                class="bold"
              >
                Cantidad: 
                2
              </p>
            </div>
            <div
              class="productTotal"
            >
              <p
                class="bold"
              >
                $
                2.000
              </p>
              <div
                class="addRemoveButtons"
              >
                <button>
                  -
                </button>
                <p>
                  2
                </p>
                <button>
                  +
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="totals"
        >
          <div
            class="labelAmount"
          >
            <p>
              Subtotal de productos
            </p>
            <p>
              $
              2.000
            </p>
          </div>
          <div
            class="labelAmount total"
          >
            <p>
              Total a pagar
            </p>
            <p>
              $
              2.000
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
